## homework 2 - data wrangling with pandas

1. [5 pts] Which NBA player scored the most points in a single season? What seasons/year was this?

In [5]:
import pandas as pd
import numpy as np

In [6]:
df = pd.read_csv("Seasons_Stats_NBA.csv")

In [7]:
df.head()

Year             Player Pos   Age   Tm     G    GS      MP   PER  \
24686  2017.0        Cody Zeller  PF  24.0  CHO  62.0  58.0  1725.0  16.7   
24687  2017.0       Tyler Zeller   C  27.0  BOS  51.0   5.0   525.0  13.0   
24688  2017.0  Stephen Zimmerman   C  20.0  ORL  19.0   0.0   108.0   7.3   
24689  2017.0        Paul Zipser  SF  22.0  CHI  44.0  18.0   843.0   6.9   
24690  2017.0        Ivica Zubac   C  19.0  LAL  38.0  11.0   609.0  17.0   

         TS%  ...    FT%    ORB    DRB    TRB   AST   STL   BLK   TOV     PF  \
24686  0.604  ...  0.679  135.0  270.0  405.0  99.0  62.0  58.0  65.0  189.0   
24687  0.508  ...  0.564   43.0   81.0  124.0  42.0   7.0  21.0  20.0   61.0   
24688  0.346  ...  0.600   11.0   24.0   35.0   4.0   2.0   5.0   3.0   17.0   
24689  0.503  ...  0.775   15.0  110.0  125.0  36.0  15.0  16.0  40.0   78.0   
24690  0.547  ...  0.653   41.0  118.0  159.0  30.0  14.0  33.0  30.0   66.0   

         PTS  
24686  639.0  
24687  178.0  
24688   23.0  
24689  240.0  
24690  284.0  

[5 rows x 52 columns]

In [10]:
df_sorted = df.sort_values("PTS", ascending = False)

In [11]:
df_sorted.head()

Year             Player Pos   Age   Tm     G    GS      MP   PER  \
1706  1962.0  Wilt Chamberlain*   C  25.0  PHW  80.0   NaN  3882.0  31.7   
1827  1963.0  Wilt Chamberlain*   C  26.0  SFW  80.0   NaN  3806.0  31.8   
8469  1987.0    Michael Jordan*  SG  23.0  CHI  82.0  82.0  3281.0  29.8   
1593  1961.0  Wilt Chamberlain*   C  24.0  PHW  79.0   NaN  3773.0  27.8   
1962  1964.0  Wilt Chamberlain*   C  27.0  SFW  80.0   NaN  3689.0  31.6   

        TS%  ...    FT%    ORB    DRB     TRB    AST    STL    BLK    TOV  \
1706  0.536  ...  0.613    NaN    NaN  2052.0  192.0    NaN    NaN    NaN   
1827  0.550  ...  0.593    NaN    NaN  1946.0  275.0    NaN    NaN    NaN   
8469  0.562  ...  0.857  166.0  264.0   430.0  377.0  236.0  125.0  272.0   
1593  0.519  ...  0.504    NaN    NaN  2149.0  148.0    NaN    NaN    NaN   
1962  0.537  ...  0.531    NaN    NaN  1787.0  403.0    NaN    NaN    NaN   

         PF     PTS  
1706  123.0  4029.0  
1827  136.0  3586.0  
8469  237.0  3041.0  
1593  130.0  3033.0  
1962  182.0  2948.0  

[5 rows x 52 columns]

In [12]:
df_sorted.tail()

Year Player  Pos  Age   Tm   G  GS  MP  PER  TS%  ...  FT%  ORB  DRB  \
21678   NaN    NaN  NaN  NaN  NaN NaN NaN NaN  NaN  NaN  ...  NaN  NaN  NaN   
22252   NaN    NaN  NaN  NaN  NaN NaN NaN NaN  NaN  NaN  ...  NaN  NaN  NaN   
22864   NaN    NaN  NaN  NaN  NaN NaN NaN NaN  NaN  NaN  ...  NaN  NaN  NaN   
23516   NaN    NaN  NaN  NaN  NaN NaN NaN NaN  NaN  NaN  ...  NaN  NaN  NaN   
24095   NaN    NaN  NaN  NaN  NaN NaN NaN NaN  NaN  NaN  ...  NaN  NaN  NaN   

       TRB  AST  STL  BLK  TOV  PF  PTS  
21678  NaN  NaN  NaN  NaN  NaN NaN  NaN  
22252  NaN  NaN  NaN  NaN  NaN NaN  NaN  
22864  NaN  NaN  NaN  NaN  NaN NaN  NaN  
23516  NaN  NaN  NaN  NaN  NaN NaN  NaN  
24095  NaN  NaN  NaN  NaN  NaN NaN  NaN  

[5 rows x 52 columns]

The NBA player who scored the most points in a single season is **Wilt Chamberlain**, during the **1962**, with a total of 4029 points.

2. [5 pts] Which NBA player scored the most points in 1991?

In [16]:
df_1991 = df_sorted[df_sorted["Year"] == 1991]

In [17]:
df_1991.head()

Year           Player Pos   Age   Tm     G    GS      MP   PER  \
10212  1991.0  Michael Jordan*  SG  27.0  CHI  82.0  82.0  3034.0  31.6   
10251  1991.0     Karl Malone*  PF  27.0  UTA  82.0  82.0  3302.0  24.8   
10132  1991.0   Patrick Ewing*   C  28.0  NYK  81.0  81.0  3104.0  23.7   
10280  1991.0    Chris Mullin*  SF  27.0  GSW  82.0  82.0  3315.0  21.4   
10336  1991.0  David Robinson*   C  25.0  SAS  82.0  81.0  3095.0  27.4   

         TS%  ...    FT%    ORB    DRB     TRB    AST    STL    BLK    TOV  \
10212  0.605  ...  0.851  118.0  374.0   492.0  453.0  223.0   83.0  202.0   
10251  0.596  ...  0.770  236.0  731.0   967.0  270.0   89.0   79.0  244.0   
10132  0.561  ...  0.745  194.0  711.0   905.0  244.0   80.0  258.0  291.0   
10280  0.618  ...  0.884  141.0  302.0   443.0  329.0  173.0   63.0  245.0   
10336  0.615  ...  0.762  335.0  728.0  1063.0  208.0  127.0  320.0  270.0   

          PF     PTS  
10212  229.0  2580.0  
10251  268.0  2382.0  
10132  287.0  2154.0  
10280  176.0  2107.0  
10336  264.0  2101.0  

[5 rows x 52 columns]

In [18]:
df_1991.tail()

Year         Player Pos   Age   Tm    G   GS    MP  PER    TS%  ...  \
10224  1991.0    Bart Kofoed  SG  26.0  GSW  5.0  0.0  21.0  0.7  0.266  ...   
10105  1991.0  Byron Dinkins  PG  23.0  IND  2.0  0.0   5.0  7.9  1.000  ...   
10385  1991.0     Jim Thomas  SG  30.0  MIN  3.0  0.0  14.0  0.7  0.250  ...   
10235  1991.0   Gary Leonard   C  23.0  ATL  4.0  0.0   9.0  8.7  0.568  ...   
10269  1991.0  Mark McNamara   C  31.0  ORL  2.0  0.0  13.0  1.0  0.000  ...   

       FT%  ORB  DRB  TRB  AST  STL  BLK  TOV   PF  PTS  
10224  0.5  2.0  1.0  3.0  4.0  0.0  0.0  2.0  4.0  3.0  
10105  NaN  0.0  1.0  1.0  0.0  0.0  0.0  0.0  3.0  2.0  
10385  NaN  0.0  0.0  0.0  1.0  1.0  0.0  1.0  0.0  2.0  
10235  0.5  0.0  2.0  2.0  0.0  0.0  1.0  0.0  2.0  2.0  
10269  NaN  0.0  4.0  4.0  0.0  0.0  0.0  0.0  1.0  0.0  

[5 rows x 52 columns]

**Michael Jordan** scored the most point in 1991.

3. [10 pts] Which player had the best free throw percentage ("FT%") from the year 2000 to the most recent year in the data? If there are multiple players who have tied for the best percentage, then sort on the column with player names and submit your answer as a pd.DataFrame object.

In [20]:
df_FT = df.sort_values("FT%", ascending = False)

In [21]:
df_FT_2000 = df_FT[df_FT["Year"] >= 2000]

In [22]:
df_FT_2000.head()

Year           Player Pos   Age   Tm     G   GS     MP   PER    TS%  \
22501  2014.0  Justin Hamilton   C  23.0  MIA   7.0  0.0   68.0  12.6  0.586   
17226  2005.0   Kaniel Dickens  SF  26.0  NJN  11.0  0.0   61.0   6.7  0.437   
19801  2009.0      Kareem Rush  SG  28.0  PHI  25.0  1.0  199.0   6.3  0.452   
22547  2014.0     Damion James  SF  26.0  SAS   5.0  1.0   50.0   4.1  0.304   
22553  2014.0   Othyus Jeffers  SG  28.0  MIN   2.0  0.0   13.0   1.7  0.532   

       ...  FT%  ORB   DRB   TRB   AST  STL  BLK   TOV    PF   PTS  
22501  ...  1.0  4.0   3.0   7.0   0.0  4.0  0.0   3.0   8.0  26.0  
17226  ...  1.0  3.0   6.0   9.0   1.0  2.0  1.0   2.0   5.0  13.0  
19801  ...  1.0  0.0  15.0  15.0  15.0  5.0  1.0  10.0  10.0  54.0  
22547  ...  1.0  1.0  11.0  12.0   3.0  0.0  1.0   1.0   3.0   6.0  
22553  ...  1.0  0.0   3.0   3.0   0.0  0.0  0.0   1.0   1.0   2.0  

[5 rows x 52 columns]

In [24]:
best_FT_Players = df_FT_2000[df_FT_2000["FT%"] == 1].sort_values("Player")

In [25]:
best_FT_Players.head()

Year           Player Pos   Age   Tm     G   GS     MP  PER    TS%  \
17990  2006.0      Aaron Miles  PG  22.0  GSW  19.0  0.0  118.0  6.0  0.405   
16181  2003.0  Adam Harrington  SG  22.0  DAL  13.0  0.0   37.0  1.4  0.308   
22787  2014.0    Adonis Thomas  SF  20.0  TOT   6.0  1.0   37.0  7.3  0.485   
22788  2014.0    Adonis Thomas  SF  20.0  ORL   4.0  0.0   24.0  5.8  0.371   
18961  2008.0   Adrian Griffin  SF  33.0  SEA  13.0  0.0   85.0  9.3  0.415   

       ...  FT%  ORB   DRB   TRB   AST  STL  BLK   TOV    PF   PTS  
17990  ...  1.0  0.0  14.0  14.0  24.0  4.0  1.0  11.0  13.0  16.0  
16181  ...  1.0  0.0   2.0   2.0   2.0  1.0  1.0   2.0   2.0  11.0  
22787  ...  1.0  0.0   3.0   3.0   3.0  0.0  0.0   1.0   6.0  14.0  
22788  ...  1.0  0.0   3.0   3.0   2.0  0.0  0.0   0.0   2.0   7.0  
18961  ...  1.0  9.0  13.0  22.0   5.0  5.0  1.0   5.0   8.0  14.0  

[5 rows x 52 columns]

In [26]:
best_FT_Players.tail()

Year           Player Pos   Age   Tm     G   GS     MP   PER    TS%  \
16906  2004.0    Wesley Person  SG  32.0  ATL   9.0  0.0  132.0  10.6  0.480   
18271  2007.0     Will Blalock  PG  23.0  DET  14.0  0.0  166.0   4.5  0.383   
17572  2005.0      Yuta Tabuse  PG  24.0  PHO   4.0  0.0   17.0  15.6  0.451   
17167  2005.0  Zarko Cabarkapa  PF  23.0  PHO   3.0  0.0   11.0  32.4  0.605   
16176  2003.0  Zendon Hamilton   C  27.0  TOR   3.0  0.0   12.0  17.7  0.510   

       ...  FT%  ORB   DRB   TRB   AST  STL  BLK   TOV   PF   PTS  
16906  ...  1.0  2.0  23.0  25.0   5.0  3.0  1.0   4.0  5.0  40.0  
18271  ...  1.0  6.0   9.0  15.0  17.0  5.0  0.0  12.0  9.0  25.0  
17572  ...  1.0  2.0   2.0   4.0   3.0  0.0  0.0   1.0  1.0   7.0  
17167  ...  1.0  2.0   1.0   3.0   0.0  0.0  0.0   0.0  0.0   9.0  
16176  ...  1.0  1.0   3.0   4.0   0.0  1.0  0.0   1.0  2.0   6.0  

[5 rows x 52 columns]

answer: **best_FT_Players**

4. [10 pts] Among those with at least 50 attempts ('FTA'), which player had the best free throw percentage ("FT%") from the year 2000 to the most recent year in the data?

In [27]:
df_filtered = df[(df["FTA"] >= 50) & (df["Year"] >= 2000)]

In [28]:
df_filtered = df_filtered.sort_values("FT%", ascending = False)

In [29]:
df_filtered.head()

Year           Player Pos   Age   Tm     G    GS      MP   PER  \
15565  2002.0  Terrell Brandon  PG  31.0  MIN  32.0  28.0   962.0  21.2   
19434  2009.0    Jose Calderon  PG  27.0  TOR  68.0  68.0  2333.0  18.8   
22484  2014.0    Danny Granger  SF  30.0  IND  29.0   2.0   653.0  10.3   
22671  2014.0        Gary Neal  SG  29.0  CHA  22.0   1.0   507.0  14.0   
19350  2009.0        Ray Allen  SG  33.0  BOS  79.0  79.0  2876.0  17.3   

         TS%  ...    FT%   ORB    DRB    TRB    AST   STL   BLK    TOV     PF  \
15565  0.494  ...  0.988  17.0   76.0   93.0  264.0  52.0   6.0   43.0   49.0   
19434  0.613  ...  0.981  16.0  178.0  194.0  607.0  74.0   7.0  143.0  114.0   
22484  0.491  ...  0.962  24.0   81.0  105.0   33.0   9.0  13.0   36.0   41.0   
22671  0.571  ...  0.961   2.0   38.0   40.0   37.0  11.0   0.0   25.0   39.0   
19350  0.624  ...  0.952  66.0  212.0  278.0  218.0  69.0  13.0  134.0  157.0   

          PTS  
15565   397.0  
19434   873.0  
22484   242.0  
22671   247.0  
19350  1434.0  

[5 rows x 52 columns]

In [30]:
df_filtered.tail()

Year           Player Pos   Age   Tm     G    GS      MP   PER  \
14837  2000.0   Olden Polynice   C  35.0  UTA  82.0  79.0  1819.0  10.4   
17715  2006.0  Andris Biedrins   C  19.0  GSW  68.0   2.0  1000.0  13.6   
23225  2015.0      Ian Mahinmi   C  28.0  IND  61.0   6.0  1146.0  10.8   
23026  2015.0      Joey Dorsey  PF  31.0  HOU  69.0  17.0   854.0  11.1   
15368  2001.0   Olden Polynice   C  36.0  UTA  81.0  79.0  1619.0   9.9   

         TS%  ...    FT%    ORB    DRB    TRB   AST   STL   BLK   TOV     PF  \
14837  0.497  ...  0.311  166.0  287.0  453.0  37.0  30.0  84.0  70.0  260.0   
17715  0.595  ...  0.306  124.0  159.0  283.0  24.0  23.0  47.0  34.0  190.0   
23225  0.516  ...  0.304  105.0  251.0  356.0  33.0  30.0  46.0  59.0  172.0   
23026  0.507  ...  0.289  103.0  176.0  279.0  25.0  40.0  25.0  41.0  157.0   
15368  0.484  ...  0.262  157.0  221.0  378.0  31.0  27.0  77.0  78.0  241.0   

         PTS  
14837  435.0  
17715  258.0  
23225  265.0  
23026  184.0  
15368  429.0  

[5 rows x 52 columns]

Among those with at least 50 attempts ('FTA'), **Terrell Brandon** had the best free throw percentage ("FT%") from the year 2000 to the most recent year in the data.

5. [5 pts] Rename the variable "Pos" to "Position".

In [31]:
df = df.rename(columns={"Pos": "Position"})

In [32]:
df.head()

Year           Player Position   Age   Tm     G  GS  MP  PER    TS%  ...  \
0  1950.0  Curly Armstrong      G-F  31.0  FTW  63.0 NaN NaN  NaN  0.368  ...   
1  1950.0     Cliff Barker       SG  29.0  INO  49.0 NaN NaN  NaN  0.435  ...   
2  1950.0    Leo Barnhorst       SF  25.0  CHS  67.0 NaN NaN  NaN  0.394  ...   
3  1950.0       Ed Bartels        F  24.0  TOT  15.0 NaN NaN  NaN  0.312  ...   
4  1950.0       Ed Bartels        F  24.0  DNN  13.0 NaN NaN  NaN  0.308  ...   

     FT%  ORB  DRB  TRB    AST  STL  BLK  TOV     PF    PTS  
0  0.705  NaN  NaN  NaN  176.0  NaN  NaN  NaN  217.0  458.0  
1  0.708  NaN  NaN  NaN  109.0  NaN  NaN  NaN   99.0  279.0  
2  0.698  NaN  NaN  NaN  140.0  NaN  NaN  NaN  192.0  438.0  
3  0.559  NaN  NaN  NaN   20.0  NaN  NaN  NaN   29.0   63.0  
4  0.548  NaN  NaN  NaN   20.0  NaN  NaN  NaN   27.0   59.0  

[5 rows x 52 columns]

In [33]:
df.tail()

Year             Player Position   Age   Tm     G    GS      MP  \
24686  2017.0        Cody Zeller       PF  24.0  CHO  62.0  58.0  1725.0   
24687  2017.0       Tyler Zeller        C  27.0  BOS  51.0   5.0   525.0   
24688  2017.0  Stephen Zimmerman        C  20.0  ORL  19.0   0.0   108.0   
24689  2017.0        Paul Zipser       SF  22.0  CHI  44.0  18.0   843.0   
24690  2017.0        Ivica Zubac        C  19.0  LAL  38.0  11.0   609.0   

        PER    TS%  ...    FT%    ORB    DRB    TRB   AST   STL   BLK   TOV  \
24686  16.7  0.604  ...  0.679  135.0  270.0  405.0  99.0  62.0  58.0  65.0   
24687  13.0  0.508  ...  0.564   43.0   81.0  124.0  42.0   7.0  21.0  20.0   
24688   7.3  0.346  ...  0.600   11.0   24.0   35.0   4.0   2.0   5.0   3.0   
24689   6.9  0.503  ...  0.775   15.0  110.0  125.0  36.0  15.0  16.0  40.0   
24690  17.0  0.547  ...  0.653   41.0  118.0  159.0  30.0  14.0  33.0  30.0   

          PF    PTS  
24686  189.0  639.0  
24687   61.0  178.0  
24688   17.0   23.0  
24689   78.0  240.0  
24690   66.0  284.0  

[5 rows x 52 columns]

6. [10 pts] Use this variable to create two variables that are called "first_position" and "second_position". Hint: If a player plays two positions, then the "-" is used as a delimiter between their first and second positions. In these cases, you will have to split the position variable into two.

In [34]:
df[["first_position", "second_position"]] = df["Position"].str.split("-", n=1, expand=True)

In [35]:
df.head()

Year           Player Position   Age   Tm     G  GS  MP  PER    TS%  ...  \
0  1950.0  Curly Armstrong      G-F  31.0  FTW  63.0 NaN NaN  NaN  0.368  ...   
1  1950.0     Cliff Barker       SG  29.0  INO  49.0 NaN NaN  NaN  0.435  ...   
2  1950.0    Leo Barnhorst       SF  25.0  CHS  67.0 NaN NaN  NaN  0.394  ...   
3  1950.0       Ed Bartels        F  24.0  TOT  15.0 NaN NaN  NaN  0.312  ...   
4  1950.0       Ed Bartels        F  24.0  DNN  13.0 NaN NaN  NaN  0.308  ...   

   DRB  TRB    AST  STL  BLK  TOV     PF    PTS  first_position  \
0  NaN  NaN  176.0  NaN  NaN  NaN  217.0  458.0               G   
1  NaN  NaN  109.0  NaN  NaN  NaN   99.0  279.0              SG   
2  NaN  NaN  140.0  NaN  NaN  NaN  192.0  438.0              SF   
3  NaN  NaN   20.0  NaN  NaN  NaN   29.0   63.0               F   
4  NaN  NaN   20.0  NaN  NaN  NaN   27.0   59.0               F   

   second_position  
0                F  
1             None  
2             None  
3             None  
4             None  

[5 rows x 54 columns]

In [36]:
df.tail()

Year             Player Position   Age   Tm     G    GS      MP  \
24686  2017.0        Cody Zeller       PF  24.0  CHO  62.0  58.0  1725.0   
24687  2017.0       Tyler Zeller        C  27.0  BOS  51.0   5.0   525.0   
24688  2017.0  Stephen Zimmerman        C  20.0  ORL  19.0   0.0   108.0   
24689  2017.0        Paul Zipser       SF  22.0  CHI  44.0  18.0   843.0   
24690  2017.0        Ivica Zubac        C  19.0  LAL  38.0  11.0   609.0   

        PER    TS%  ...    DRB    TRB   AST   STL   BLK   TOV     PF    PTS  \
24686  16.7  0.604  ...  270.0  405.0  99.0  62.0  58.0  65.0  189.0  639.0   
24687  13.0  0.508  ...   81.0  124.0  42.0   7.0  21.0  20.0   61.0  178.0   
24688   7.3  0.346  ...   24.0   35.0   4.0   2.0   5.0   3.0   17.0   23.0   
24689   6.9  0.503  ...  110.0  125.0  36.0  15.0  16.0  40.0   78.0  240.0   
24690  17.0  0.547  ...  118.0  159.0  30.0  14.0  33.0  30.0   66.0  284.0   

       first_position  second_position  
24686              PF             None  
24687               C             None  
24688               C             None  
24689              SF             None  
24690               C             None  

[5 rows x 54 columns]

7. [5 pts] Combine these two variables back into a single variable called "position_united". Make sure to separate the two positions by "-", as the original data field did. If there is no second position, then there will only be a single position. Again, this is similar to how the original variable was coded. Present a frequency table of "position_united".

In [37]:
df["position_united"] = df["first_position"]

In [38]:
df.loc[df["second_position"].notna(), "position_united"] = (
    df["first_position"] + "-" + df["second_position"]
)

In [42]:
position_united = df["position_united"].value_counts().reset_index()
position_united.columns = ["position_united", "frequency"]
position_frequency.head()

position_united  frequency
0              PF       4966
1              SG       4811
2               C       4759
3              SF       4699
4              PG       4648

In [43]:
position_united.tail()

position_united  frequency
18           SF-PF         21
19           SG-PF          3
20            C-SF          2
21           SF-PG          1
22           PG-SF          1

8. [10 pts] Create two new datasets.

a new dataset from the original dataset that includes all data except the age variable (be sure to give this dataset a new name). 
a new dataset from the original dataset that includes the year, the player name, and age.
add a new column to both datasets called mergeid that includes a sequence of numbers beginning with a 1 in the first row of the data and ending with the total number of rows in the last row of the data

In [44]:
df_no_age = df.drop(columns=["Age"])

In [45]:
df_no_age.head()

Year           Player Position   Tm     G  GS  MP  PER    TS%  3PAr  ...  \
0  1950.0  Curly Armstrong      G-F  FTW  63.0 NaN NaN  NaN  0.368   NaN  ...   
1  1950.0     Cliff Barker       SG  INO  49.0 NaN NaN  NaN  0.435   NaN  ...   
2  1950.0    Leo Barnhorst       SF  CHS  67.0 NaN NaN  NaN  0.394   NaN  ...   
3  1950.0       Ed Bartels        F  TOT  15.0 NaN NaN  NaN  0.312   NaN  ...   
4  1950.0       Ed Bartels        F  DNN  13.0 NaN NaN  NaN  0.308   NaN  ...   

   TRB    AST  STL  BLK  TOV     PF    PTS  first_position  second_position  \
0  NaN  176.0  NaN  NaN  NaN  217.0  458.0               G                F   
1  NaN  109.0  NaN  NaN  NaN   99.0  279.0              SG             None   
2  NaN  140.0  NaN  NaN  NaN  192.0  438.0              SF             None   
3  NaN   20.0  NaN  NaN  NaN   29.0   63.0               F             None   
4  NaN   20.0  NaN  NaN  NaN   27.0   59.0               F             None   

   position_united  
0              G-F  
1               SG  
2               SF  
3                F  
4                F  

[5 rows x 54 columns]

In [46]:
df_no_age.tail()

Year             Player Position   Tm     G    GS      MP   PER  \
24686  2017.0        Cody Zeller       PF  CHO  62.0  58.0  1725.0  16.7   
24687  2017.0       Tyler Zeller        C  BOS  51.0   5.0   525.0  13.0   
24688  2017.0  Stephen Zimmerman        C  ORL  19.0   0.0   108.0   7.3   
24689  2017.0        Paul Zipser       SF  CHI  44.0  18.0   843.0   6.9   
24690  2017.0        Ivica Zubac        C  LAL  38.0  11.0   609.0  17.0   

         TS%   3PAr  ...    TRB   AST   STL   BLK   TOV     PF    PTS  \
24686  0.604  0.002  ...  405.0  99.0  62.0  58.0  65.0  189.0  639.0   
24687  0.508  0.006  ...  124.0  42.0   7.0  21.0  20.0   61.0  178.0   
24688  0.346  0.000  ...   35.0   4.0   2.0   5.0   3.0   17.0   23.0   
24689  0.503  0.448  ...  125.0  36.0  15.0  16.0  40.0   78.0  240.0   
24690  0.547  0.013  ...  159.0  30.0  14.0  33.0  30.0   66.0  284.0   

       first_position  second_position  position_united  
24686              PF             None               PF  
24687               C             None                C  
24688               C             None                C  
24689              SF             None               SF  
24690               C             None                C  

[5 rows x 54 columns]

In [47]:
df_year_player_age = df[["Year", "Player", "Age"]]

In [48]:
df_year_player_age.head()

Year           Player   Age
0  1950.0  Curly Armstrong  31.0
1  1950.0     Cliff Barker  29.0
2  1950.0    Leo Barnhorst  25.0
3  1950.0       Ed Bartels  24.0
4  1950.0       Ed Bartels  24.0

In [49]:
df_year_player_age.tail()

Year             Player   Age
24686  2017.0        Cody Zeller  24.0
24687  2017.0       Tyler Zeller  27.0
24688  2017.0  Stephen Zimmerman  20.0
24689  2017.0        Paul Zipser  22.0
24690  2017.0        Ivica Zubac  19.0

In [51]:
df_no_age.loc[:, "mergeid"] = range(1, len(df_no_age) + 1)
df_year_player_age.loc[:, "mergeid"] = range(1, len(df_year_player_age) + 1)

In [52]:
df_no_age.head()

Year           Player Position   Tm     G  GS  MP  PER    TS%  3PAr  ...  \
0  1950.0  Curly Armstrong      G-F  FTW  63.0 NaN NaN  NaN  0.368   NaN  ...   
1  1950.0     Cliff Barker       SG  INO  49.0 NaN NaN  NaN  0.435   NaN  ...   
2  1950.0    Leo Barnhorst       SF  CHS  67.0 NaN NaN  NaN  0.394   NaN  ...   
3  1950.0       Ed Bartels        F  TOT  15.0 NaN NaN  NaN  0.312   NaN  ...   
4  1950.0       Ed Bartels        F  DNN  13.0 NaN NaN  NaN  0.308   NaN  ...   

     AST  STL  BLK  TOV     PF    PTS  first_position  second_position  \
0  176.0  NaN  NaN  NaN  217.0  458.0               G                F   
1  109.0  NaN  NaN  NaN   99.0  279.0              SG             None   
2  140.0  NaN  NaN  NaN  192.0  438.0              SF             None   
3   20.0  NaN  NaN  NaN   29.0   63.0               F             None   
4   20.0  NaN  NaN  NaN   27.0   59.0               F             None   

   position_united  mergeid  
0              G-F        1  
1               SG        2  
2               SF        3  
3                F        4  
4                F        5  

[5 rows x 55 columns]

In [53]:
df_no_age.tail()

Year             Player Position   Tm     G    GS      MP   PER  \
24686  2017.0        Cody Zeller       PF  CHO  62.0  58.0  1725.0  16.7   
24687  2017.0       Tyler Zeller        C  BOS  51.0   5.0   525.0  13.0   
24688  2017.0  Stephen Zimmerman        C  ORL  19.0   0.0   108.0   7.3   
24689  2017.0        Paul Zipser       SF  CHI  44.0  18.0   843.0   6.9   
24690  2017.0        Ivica Zubac        C  LAL  38.0  11.0   609.0  17.0   

         TS%   3PAr  ...   AST   STL   BLK   TOV     PF    PTS  \
24686  0.604  0.002  ...  99.0  62.0  58.0  65.0  189.0  639.0   
24687  0.508  0.006  ...  42.0   7.0  21.0  20.0   61.0  178.0   
24688  0.346  0.000  ...   4.0   2.0   5.0   3.0   17.0   23.0   
24689  0.503  0.448  ...  36.0  15.0  16.0  40.0   78.0  240.0   
24690  0.547  0.013  ...  30.0  14.0  33.0  30.0   66.0  284.0   

       first_position  second_position  position_united  mergeid  
24686              PF             None               PF    24687  
24687               C             None                C    24688  
24688               C             None                C    24689  
24689              SF             None               SF    24690  
24690               C             None                C    24691  

[5 rows x 55 columns]

In [54]:
df_year_player_age.head()

Year           Player   Age  mergeid
0  1950.0  Curly Armstrong  31.0        1
1  1950.0     Cliff Barker  29.0        2
2  1950.0    Leo Barnhorst  25.0        3
3  1950.0       Ed Bartels  24.0        4
4  1950.0       Ed Bartels  24.0        5

In [55]:
df_year_player_age.tail()

Year             Player   Age  mergeid
24686  2017.0        Cody Zeller  24.0    24687
24687  2017.0       Tyler Zeller  27.0    24688
24688  2017.0  Stephen Zimmerman  20.0    24689
24689  2017.0        Paul Zipser  22.0    24690
24690  2017.0        Ivica Zubac  19.0    24691

9. [10 pts] Join the two datasets from question (8) together to recreate the original dataset plus the new merge id.

In [56]:
df_recreated = pd.merge(
    df_no_age,
    df_year_player_age,
    on="mergeid",
    how="inner"
)

In [57]:
df_recreated.head()

Year_x         Player_x Position   Tm     G  GS  MP  PER    TS%  3PAr  ...  \
0  1950.0  Curly Armstrong      G-F  FTW  63.0 NaN NaN  NaN  0.368   NaN  ...   
1  1950.0     Cliff Barker       SG  INO  49.0 NaN NaN  NaN  0.435   NaN  ...   
2  1950.0    Leo Barnhorst       SF  CHS  67.0 NaN NaN  NaN  0.394   NaN  ...   
3  1950.0       Ed Bartels        F  TOT  15.0 NaN NaN  NaN  0.312   NaN  ...   
4  1950.0       Ed Bartels        F  DNN  13.0 NaN NaN  NaN  0.308   NaN  ...   

   TOV     PF    PTS  first_position  second_position  position_united  \
0  NaN  217.0  458.0               G                F              G-F   
1  NaN   99.0  279.0              SG             None               SG   
2  NaN  192.0  438.0              SF             None               SF   
3  NaN   29.0   63.0               F             None                F   
4  NaN   27.0   59.0               F             None                F   

   mergeid  Year_y         Player_y   Age  
0        1  1950.0  Curly Armstrong  31.0  
1        2  1950.0     Cliff Barker  29.0  
2        3  1950.0    Leo Barnhorst  25.0  
3        4  1950.0       Ed Bartels  24.0  
4        5  1950.0       Ed Bartels  24.0  

[5 rows x 58 columns]

In [58]:
df_recreated.tail()

Year_x           Player_x Position   Tm     G    GS      MP   PER  \
24686  2017.0        Cody Zeller       PF  CHO  62.0  58.0  1725.0  16.7   
24687  2017.0       Tyler Zeller        C  BOS  51.0   5.0   525.0  13.0   
24688  2017.0  Stephen Zimmerman        C  ORL  19.0   0.0   108.0   7.3   
24689  2017.0        Paul Zipser       SF  CHI  44.0  18.0   843.0   6.9   
24690  2017.0        Ivica Zubac        C  LAL  38.0  11.0   609.0  17.0   

         TS%   3PAr  ...   TOV     PF    PTS  first_position  second_position  \
24686  0.604  0.002  ...  65.0  189.0  639.0              PF             None   
24687  0.508  0.006  ...  20.0   61.0  178.0               C             None   
24688  0.346  0.000  ...   3.0   17.0   23.0               C             None   
24689  0.503  0.448  ...  40.0   78.0  240.0              SF             None   
24690  0.547  0.013  ...  30.0   66.0  284.0               C             None   

       position_united  mergeid  Year_y           Player_y   Age  
24686               PF    24687  2017.0        Cody Zeller  24.0  
24687                C    24688  2017.0       Tyler Zeller  27.0  
24688                C    24689  2017.0  Stephen Zimmerman  20.0  
24689               SF    24690  2017.0        Paul Zipser  22.0  
24690                C    24691  2017.0        Ivica Zubac  19.0  

[5 rows x 58 columns]

10. [10 pts] Subset the original dataset to the years 1995-1997. Group the data by team name and year and then summarize the average number of points per team. Arrange from most to least points. (Hint: Your answer should have a shape of 88 x 3, since we are looking at the average points scored over the 1995-1997 seasons. Note that 2 of these columns could be indexes, in which case Pandas will tell you your data is 88 rows x 1 columns. Finally, also note that the data for VAN and TOR are missing in 1995 as these teams did not exist in that year. So your answer could also be 90 x 3, if you include missing/empty data in these rows).

In [59]:
df_9597 = df[(df["Year"] >= 1995) & (df["Year"] <= 1997)]

In [61]:
avg_points_team_year = (
    df_9597
    .groupby(["Tm", "Year"])["PTS"]
    .mean()
)

In [62]:
avg_points_team_year = avg_points_team_year.reset_index(name="avg_points")

In [63]:
avg_points_team_year = avg_points_team_year.sort_values(
    by="avg_points",
    ascending=False
)

In [65]:
avg_points_team_year.shape

(88, 3)

In [66]:
avg_points_team_year.head()

Tm    Year  avg_points
73  SEA  1996.0  659.384615
72  SEA  1995.0  646.785714
47  MIN  1997.0  606.307692
54  ORL  1995.0  606.066667
60  PHO  1995.0  604.866667

In [67]:
avg_points_team_year.tail()

Tm    Year  avg_points
20  DEN  1997.0  348.695652
50  NJN  1997.0  346.695652
58  PHI  1996.0  322.750000
79  TOT  1997.0  322.380952
17  DAL  1997.0  275.222222

11. [10 pts] Similar to Q10, start by subsetting the original dataset to the years 1995-1997. Then create a wide dataset that keeps each year in its own row, but spreads team names to multiple columns with each column delineating points per team in each of the 1995-1997 seasons (Hint: Your final answer should have a shape of 3 x 30).

In [68]:
df_9597 = df[(df["Year"] >= 1995) & (df["Year"] <= 1997)]

In [69]:
avg_points_team_year = (
    df_9597
    .groupby(["Year", "Tm"])["PTS"]
    .mean()
    .reset_index()
)

In [70]:
wide_points = avg_points_team_year.pivot(
    index="Year",
    columns="Tm",
    values="PTS"
)

In [72]:
wide_points.shape

(3, 30)

In [73]:
wide_points.head()

Tm             ATL         BOS         CHH         CHI         CLE  \
Year                                                                 
1995.0  440.055556  495.764706  515.562500  520.312500  494.466667   
1996.0  447.722222  471.944444  401.476190  575.000000  498.200000   
1997.0  457.294118  458.222222  450.444444  563.866667  551.769231   

Tm             DAL         DEN         DET         GSW         HOU  ...  \
Year                                                                ...   
1995.0  604.428571  488.764706  503.312500  481.500000  499.470588  ...   
1996.0  560.600000  471.352941  558.714286  490.235294  466.888889  ...   
1997.0  275.222222  348.695652  482.687500  510.687500  458.222222  ...   

Tm             PHO         POR         SAC         SAS         SEA  \
Year                                                                 
1995.0  604.866667  528.187500  537.066667  546.375000  646.785714   
1996.0  503.058824  543.000000  510.187500  565.133333  659.384615   
1997.0  366.565217  427.052632  494.250000  370.900000  517.125000   

Tm             TOR         TOT         UTA       VAN         WSB  
Year                                                              
1995.0         NaN  488.681818  581.733333       NaN  515.125000  
1996.0  420.736842  419.767857  600.285714  409.0000  400.380952  
1997.0  372.809524  322.380952  563.600000  457.0625  543.133333  

[3 rows x 30 columns]

In [74]:
wide_points.tail()

Tm             ATL         BOS         CHH         CHI         CLE  \
Year                                                                 
1995.0  440.055556  495.764706  515.562500  520.312500  494.466667   
1996.0  447.722222  471.944444  401.476190  575.000000  498.200000   
1997.0  457.294118  458.222222  450.444444  563.866667  551.769231   

Tm             DAL         DEN         DET         GSW         HOU  ...  \
Year                                                                ...   
1995.0  604.428571  488.764706  503.312500  481.500000  499.470588  ...   
1996.0  560.600000  471.352941  558.714286  490.235294  466.888889  ...   
1997.0  275.222222  348.695652  482.687500  510.687500  458.222222  ...   

Tm             PHO         POR         SAC         SAS         SEA  \
Year                                                                 
1995.0  604.866667  528.187500  537.066667  546.375000  646.785714   
1996.0  503.058824  543.000000  510.187500  565.133333  659.384615   
1997.0  366.565217  427.052632  494.250000  370.900000  517.125000   

Tm             TOR         TOT         UTA       VAN         WSB  
Year                                                              
1995.0         NaN  488.681818  581.733333       NaN  515.125000  
1996.0  420.736842  419.767857  600.285714  409.0000  400.380952  
1997.0  372.809524  322.380952  563.600000  457.0625  543.133333  

[3 rows x 30 columns]

12. [10 pts] Now return the data from #11 back into a long (tidy) format by moving Team (TM), Points and Year into their own columns (Hint: Your data shape should be 90 x 3).

In [75]:
wide_points_reset = wide_points.reset_index()

In [76]:
long_points = wide_points_reset.melt(
    id_vars="Year",
    var_name="Tm",
    value_name="Points"
)

In [77]:
long_points.shape

(90, 3)

In [78]:
long_points.head()

Year   Tm      Points
0  1995.0  ATL  440.055556
1  1996.0  ATL  447.722222
2  1997.0  ATL  457.294118
3  1995.0  BOS  495.764706
4  1996.0  BOS  471.944444

In [79]:
long_points.tail()

Year   Tm      Points
85  1996.0  VAN  409.000000
86  1997.0  VAN  457.062500
87  1995.0  WSB  515.125000
88  1996.0  WSB  400.380952
89  1997.0  WSB  543.133333